django/core/validators.py

```diff
@@ -86,7 +86,14 @@ class URLValidator(RegexValidator):
     hostname_re = r'[a-z' + ul + r'0-9](?:[a-z' + ul + r'0-9-]{0,61}[a-z' + ul + r'0-9])?'
     # Max length for domain name labels is 63 characters per RFC 1034 sec. 3.1
     domain_re = r'(?:\.(?!-)[a-z' + ul + r'0-9-]{1,63}(?<!-))*'
-    tld_re = r'\.(?:[a-z' + ul + r']{2,63}|xn--[a-z0-9]{1,59})\.?'
+    tld_re = (
+        '\.'                                # dot
+        '(?!-)'                             # can't start with a dash
+        '(?:[a-z-' + ul + ']{2,63}'         # domain label
+        '|xn--[a-z0-9]{1,59})'              # or punycode label
+        '\.?'                               # may have a trailing dot
+        '(?<!-)'                            # can't end with a dash
```

Comment:
The optional trailing dot may "protect" an ending dash -- `xn--.` passes this. Is this intended? If not, switch the last two lines.